app/src/main/java/at/tacticaldevc/oat/utils/Prefs.java

```diff
@@ -27,16 +34,125 @@
     private final static String DOCUMENT_NAME_ENABLED_FEATURES = "oat-enabled-features";
     private final static String DOCUMENT_NAME_ACCEPTED_CONDITIONS = "oat-accepted-conditions";
     private final static String KEY_TRUSTED_CONTACTS = "trusted-contacts";
+    private final static String KEY_COMMAND_PASSWORD = "password";
+    private final static String KEY_COMMAND_PASSWORD_SALT = "pwdsalt";
+    private final static String KEY_COMMAND_TRIGGER = "cmd-trigger";
+    private final static String KEY_MISSING_PERMISSIONS_TO_REQUEST_ON_STARTUP = "missing-permission";
+
+    // Basic Data
+
+    /**
+     * Updates the Password used to send commands to the App
+     *
+     * @param context  the Context of the Application
+     * @param password the new Password
+     */
+    public static void savePassword(Context context, String password) {
+        ensureNotNull(context, "Application Context");
+        ensureStringIsValid(password, "new User Password");
+
+        SharedPreferences prefs = context.getSharedPreferences(DOCUMENT_NAME_DATA, Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = prefs.edit();
+
+        byte[] salt;
+        MessageDigest algorithm;
+        try {
+            algorithm = MessageDigest.getInstance("SHA-256");
+        } catch (NoSuchAlgorithmException ex) {
+            throw OATApplicationException.forLibraryError("MessageDigest", ex);
+        }
+
+        if (prefs.getString(KEY_COMMAND_PASSWORD_SALT, null) == null) {
+            SecureRandom rnd = new SecureRandom();
+            salt = new byte[16];
+            rnd.nextBytes(salt);
+            editor.putString(KEY_COMMAND_PASSWORD_SALT, Base64.encodeToString(salt, Base64.NO_WRAP));
+        } else {
+            salt = Base64.decode(prefs.getString(KEY_COMMAND_PASSWORD_SALT, null), Base64.NO_WRAP);
+        }
+        algorithm.update(salt);
+        String hash = Base64.encodeToString(algorithm.digest(password.getBytes()), Base64.NO_WRAP);
+
+        editor.putString(KEY_COMMAND_PASSWORD, hash);
+        editor.apply();
+    }
+
+    /**
+     * Verifies that a given Password is correct
+     *
+     * @param context         the Context of the Application
+     * @param passwordToCheck the password to be verified
+     * @return true if the password matches the password that was stored
+     */
+    public static boolean verifyApplicationPassword(Context context, String passwordToCheck) {
+        ensureNotNull(context, "Application Context");
+        try {
+            ensureStringIsValid(passwordToCheck, "password to check");
+        } catch (IllegalArgumentException ex) {
+            return false;
+        }
+
+        SharedPreferences prefs = context.getSharedPreferences(DOCUMENT_NAME_DATA, Context.MODE_PRIVATE);
+
+        String salt = prefs.getString(KEY_COMMAND_PASSWORD_SALT, "");
```

Comment:
Maybe you should do an if here for checking for a salt. Wouldn't make sense if there is a password but no hash. In that specific case we would have to do a full application reset (purge data, cache, ...)

Reply:
password is now deleted if this situation should occur.